File: 17.电话号码的字母组合/letterCombinations.go
package problem0017

var phoneMap = map[byte]string{
	'2': "abc",
	'3': "def",
	'4': "ghi",
	'5': "jkl",
	'6': "mno",
	'7': "pqrs",
	'8': "tuv",
	'9': "wxyz",
}

func letterCombinations(digits string) []string {
	var (
		result    []string
		backTrack func(path string, index int)
	)
	backTrack = func(path string, index int) {
		if len(path) == len(digits) {
			if path != "" {
				result = append(result, path)
			}
			return
		}
		options := phoneMap[digits[index]]
		for i := 0; i < len(options); i++ {
			path = path + string(options[i])
			backTrack(path, index+1)
			path = path[:len(path)-1]
		}
	}
	backTrack("", 0)
	return result
}

File: 17.电话号码的字母组合/letterCombinations.js
const map = {
	2: 'abc',
	3: 'def',
	4: 'ghi',
	5: 'jkl',
	6: 'mno',
	7: 'pqrs',
	8: 'tuv',
	9: 'wxyz',
}

function letterCombinations(digits) {
	if (digits.length === 0) {
		return []
	}
	const result = []
	const backTrack = (combine, index) => {
		if (combine.length === digits.length) {
			result.push(combine)
			return
		}
		const options = map[digits.charAt(index)]
		for (let i = 0; i < options.length; i++) {
			combine += options.charAt(i)
			backTrack(combine, index + 1)
			combine = combine.slice(0, combine.length - 1)
		}
	}
	backTrack('', 0)
	return result
}